File: Leetcode/872. Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int>arr1,arr2;
        inorder(root1,arr1);
        inorder(root2,arr2);
        if(arr1==arr2)
            return true;
        return false;
    }
    void inorder(TreeNode* root, vector<int>&arr){
        if(root!=NULL){
            inorder(root->left,arr);
            inorder(root->right,arr);
            if(root->left==NULL && root->right==NULL)
                arr.push_back(root->val);
        }
    }

};

File: Leetcode/872. Leaf-Similar Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        ArrayList<Integer> leaf1 = new ArrayList<>();;
        ArrayList<Integer> leaf2 = new ArrayList<>();
        dfs(root1, leaf1);
        dfs(root2, leaf2);
        return leaf1.equals(leaf2);
    }
    private void dfs(TreeNode root, ArrayList<Integer> leaf) {
        if(root==null)
            return;
        if(root.left==null && root.right==null) {
            leaf.add(root.val);
            return;
        }
        dfs(root.left, leaf);
        dfs(root.right, leaf);
    }
}
